### PR TITLE
UGENE-7011 Add possibility to choose amino acid translation frame in the Export Amino Translations feature in Alignment Editor

### PR DESCRIPTION
**Pull request by DmitriiSukhomlinov**
https://ugene.dev/tracker/browse/UGENE-7011

### CODE REVIEW

File: src/plugins/dna_export/src/ExportAlignmentViewItems.cpp [hunk 1]
```diff
@@ -120,14 +120,19 @@ void MSAExportContext::sl_exportNucleicMsaToAmino() {
     int len = d->exportWholeAlignment ? ma->getNumRows() : editor->getSelectionRect().height();
 
     bool convertUnknowToGaps = d->unknownAmino == ExportMSA2MSADialog::UnknownAmino::Gap;
+    int frameNum = static_cast<int>(d->translationFrame);
+    bool reverseCompement = frameNum < 0;
+    int baseOffset = (qAbs(frameNum) - 1);
```
Comment (yalgaer) [inline src/plugins/dna_export/src/ExportAlignmentViewItems.cpp]: nit: extra braces are not needed

File: src/plugins/dna_export/src/ExportMSA2MSADialog.cpp [hunk 2]
```diff
@@ -77,6 +78,22 @@ void ExportMSA2MSADialog::updateModel() {
     if (includeGaps) {
         unknownAmino = rbUseX->isChecked() ? UnknownAmino::X : UnknownAmino::Gap;
     }
+
+    if (rb1->isChecked()) {
+        translationFrame = TranslationFrame::One;
+    } else if (rb2->isChecked()) {
```
Comment (yalgaer) [inline src/plugins/dna_export/src/ExportMSA2MSADialog.cpp]: Please give a meaningful name to the radio buttons.

File: src/plugins/dna_export/src/ExportMSA2MSADialog.h [hunk 3]
```diff
@@ -44,6 +44,19 @@ class ExportMSA2MSADialog : public QDialog, Ui_ExportMSA2MSADialog {
         Gap
     };
 
+    /*!
+     * \class TranslationFrame
+     * \brief Frame to translate (reverse-complement if minus)
+     */
+    enum class TranslationFrame {
```
Comment (yalgaer) [inline src/plugins/dna_export/src/ExportMSA2MSADialog.h]: Do we want to introduce such enum in every task/dialog where translation frames are used? 
If yes: this is overdesign. If not: either add the enum to some common core header or use simple 1,2,3,-1,-2,-3 in this class as we did before in other places.

If you will decide to add the enum please use more meaningful names for the fields: FirstDirectFrame FirstReverseComplementFrame or similar.

Naming "-1" as MinusOne makes no value.

File: src/plugins/dna_export/src/DNAExportPluginTests.cpp [hunk 4]
```diff
@@ -194,14 +206,19 @@ void GTest_ExportNucleicToAminoAlignmentTask::prepare() {
     trid.replace("0", QString("%1").arg(transTable));
     trans << AppContext::getDNATranslationRegistry()->lookupTranslation(trid);
 
+    int frameNum = static_cast<int>(translationFrame);
```
Comment (yalgaer) [inline src/plugins/dna_export/src/DNAExportPluginTests.cpp]: Do we really need this cast/local var?

File: src/plugins/dna_export/src/DNAExportPluginTests.cpp [hunk 5]
```diff
@@ -194,14 +206,19 @@ void GTest_ExportNucleicToAminoAlignmentTask::prepare() {
...
+    bool reverseCompement = frameNum < 0;
+    int offset = (qAbs(frameNum) - 1);
```
Comment (yalgaer) [inline src/plugins/dna_export/src/DNAExportPluginTests.cpp]: nit: useless braces

File: src/plugins/dna_export/src/ExportTasks.cpp [hunk 6]
```diff
@@ -197,6 +203,9 @@ void ExportMSA2MSATask::run() {
             }
             resseq.replace("*", "X");
             DNASequence rs(name, resseq, aminoTT->getDstAlphabet());
+            if (reverseComplement) {
```
Comment (yalgaer) [inline src/plugins/dna_export/src/ExportTasks.cpp]: This is wrong. Reversed-only sequences have no sense/meaning.
You already did reverse-complement above and it is enough.

**Review by yalgaer — CHANGES_REQUESTED** [review]

File: src/plugins/dna_export/src/ExportAlignmentViewItems.cpp (hunk 1 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/dna_export/src/ExportAlignmentViewItems.cpp]: Fixed

File: src/plugins/dna_export/src/ExportMSA2MSADialog.h (hunk 3 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/dna_export/src/ExportMSA2MSADialog.h]: I decided to get rid from the enum

File: src/plugins/dna_export/src/DNAExportPluginTests.cpp (hunk 4 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/dna_export/src/DNAExportPluginTests.cpp]: No, the "copy-paste" victim

File: src/plugins/dna_export/src/DNAExportPluginTests.cpp (hunk 5 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/dna_export/src/DNAExportPluginTests.cpp]: Fixed

File: src/plugins/dna_export/src/ExportTasks.cpp (hunk 6 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/dna_export/src/ExportTasks.cpp]: The meaning of this is to turn from left-to-right to right-to-left representation as it's done in sequence view (the reverse-complement strand in SV is written from right to left). I can remove this if you want.

File: src/plugins/dna_export/src/ExportMSA2MSADialog.cpp (hunk 2 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/dna_export/src/ExportMSA2MSADialog.cpp]: Fixed

**Comment by DmitriiSukhomlinov** [comment]
@yalgaer check again

File: src/plugins/dna_export/src/ExportTasks.cpp (hunk 6 above)
Comment (yalgaer) [reply src/plugins/dna_export/src/ExportTasks.cpp]: Amino sequences have only 1 reading direction. Even in the sequence view if you copy the rev-compl translation you will have a 'one-way' (left-to-right) amino sequence.

I believe the reverse here is not needed and should be removed.

File: src/plugins/GUITestBase/src/runnables/ugene/plugins/dna_export/ExportMSA2MSADialogFiller.cpp [hunk 7]
```diff
@@ -66,6 +67,27 @@ void ExportMSA2MSADialogFiller::commonScenario() {
         }
     }
 
+    if (frame != 1) {
+        QString widgetName;
+        if (frame == 1) {
```
Comment (evelinabis) [inline src/plugins/GUITestBase/src/runnables/ugene/plugins/dna_export/ExportMSA2MSADialogFiller.cpp]: I see
```
if (frame! = 1) {
    if (frame == 1) {
        ...
```
and it confuses me

Comment (evelinabis) [inline src/plugins/GUITestBase/src/runnables/ugene/plugins/dna_export/ExportMSA2MSADialogFiller.cpp]: Why are you using if-else and not a switch?

File: src/plugins/dna_export/src/DNAExportPluginTests.cpp [hunk 8]
```diff
@@ -194,14 +206,18 @@ void GTest_ExportNucleicToAminoAlignmentTask::prepare() {
     trid.replace("0", QString("%1").arg(transTable));
     trans << AppContext::getDNATranslationRegistry()->lookupTranslation(trid);
 
+    bool reverseCompement = translationFrame < 0;
```
Comment (evelinabis) [inline src/plugins/dna_export/src/DNAExportPluginTests.cpp]: Typo

File: src/plugins/dna_export/src/ExportAlignmentViewItems.cpp [hunk 9]
```diff
@@ -120,14 +120,18 @@ void MSAExportContext::sl_exportNucleicMsaToAmino() {
     int len = d->exportWholeAlignment ? ma->getNumRows() : editor->getSelectionRect().height();
 
     bool convertUnknowToGaps = d->unknownAmino == ExportMSA2MSADialog::UnknownAmino::Gap;
+    bool reverseCompement = d->translationFrame < 0;
```
Comment (evelinabis) [inline src/plugins/dna_export/src/ExportAlignmentViewItems.cpp]: Same typo 1

File: src/plugins/dna_export/src/ExportProjectViewItems.cpp [hunk 10]
```diff
@@ -507,14 +507,18 @@ void ExportProjectViewItemsContoller::sl_exportNucleicAlignmentToAmino() {
     trans << AppContext::getDNATranslationRegistry()->lookupTranslation(d->translationTable);
 
     bool convertUnknowToGaps = d->unknownAmino == ExportMSA2MSADialog::UnknownAmino::Gap;
+    bool reverseCompement = d->translationFrame < 0;
```
Comment (evelinabis) [inline src/plugins/dna_export/src/ExportProjectViewItems.cpp]: Same typo 2

**Review by evelinabis — DISMISSED** [review]

File: src/plugins/GUITestBase/src/runnables/ugene/plugins/dna_export/ExportMSA2MSADialogFiller.cpp (hunk 7 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/GUITestBase/src/runnables/ugene/plugins/dna_export/ExportMSA2MSADialogFiller.cpp]: Fixed

File: src/plugins/dna_export/src/DNAExportPluginTests.cpp (hunk 8 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/dna_export/src/DNAExportPluginTests.cpp]: Fixed

File: src/plugins/dna_export/src/ExportAlignmentViewItems.cpp (hunk 9 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/dna_export/src/ExportAlignmentViewItems.cpp]: Fixed

File: src/plugins/dna_export/src/ExportProjectViewItems.cpp (hunk 10 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/dna_export/src/ExportProjectViewItems.cpp]: Fixed

File: src/plugins/dna_export/src/ExportTasks.cpp (hunk 6 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/dna_export/src/ExportTasks.cpp]: OK, fixed

**Comment by DmitriiSukhomlinov** [comment]
@yalgaer @EvelinaBiserova fixed, check again

**Review by evelinabis — APPROVED** [review]

**Comment by DmitriiSukhomlinov** [comment]
I've run the Teamcity suite [here](http://ugene-teamcity.unipro.ru:8111/ugene-teamcity/viewLog.html?buildId=212193&tab=buildResultsDiv&buildTypeId=bt153), but there is a test which is failed. This test isn't related to this feature at all (it's about shared database), so, I suggest, that it's some kind of temporary failure. You may either approve or wait for another run, which is, probably, you will have waited for at least a day.

**Review by yalgaer — APPROVED** [review]

**Pull request merged**